packages/aws-client/src/S3.ts

```diff
@@ -1,44 +1,49 @@
-const fs = require('fs');
-const isString = require('lodash/isString');
-const path = require('path');
-const pMap = require('p-map');
-const pRetry = require('p-retry');
-const pump = require('pump');
-const url = require('url');
-
-const {
+/**
+ * @module S3
+ */
+
+import fs = require('fs');
+import path = require('path');
+import pMap = require('p-map');
+import pRetry = require('p-retry');
+import pump = require('pump');
+import url = require('url');
+import { Readable, TransformOptions } from 'stream';
+
+import {
   generateChecksumFromStream,
   validateChecksumFromStream
-} = require('@cumulus/checksum');
-const {
+} from '@cumulus/checksum';
+import {
   InvalidChecksum,
   UnparsableFileLocationError
-} = require('@cumulus/errors');
-const Logger = require('@cumulus/logger');
+} from '@cumulus/errors';
+import Logger = require('@cumulus/logger');
 
-const awsServices = require('./services');
-const { inTestMode } = require('./test-utils');
-const { improveStackTrace } = require('./utils');
+import { s3 } from './services';
+import { inTestMode } from './test-utils';
+import { improveStackTrace } from './utils';
 
 const log = new Logger({ sender: 'aws-client/s3' });
 
-let S3_RATE_LIMIT = 20;
-if (inTestMode()) {
-  S3_RATE_LIMIT = 1;
-}
+const S3_RATE_LIMIT = inTestMode() ? 1 : 20;
 
 /**
  * Join strings into an S3 key without a leading slash or double slashes
  *
  * @param {...string|Array<string>} args - the strings to join
  * @returns {string} the full S3 key
+ *
+ * @static
  */
-function s3Join(...args) {
-  const tokens = Array.isArray(args[0]) ? args[0] : args;
+export function s3Join(...args: [string | string[], ...string[]]) {
```

Comment:
`args` is a string, an array of strings, or ?

Reply:
Yeah, I would love to break the backward compatibility of this function. You can call `s3Join` with either an array of strings to join, or any number of strings to be joined (rather than an array). It would be a lot simpler if it behaved like `path.join` and just took a variable number of strings rather than that _or_ an array.

This is saying that the first argument to `s3Join` can be either a string or an array of strings. Any arguments _after_ that first one must be strings. Say the word and I'll be happy to make a breaking change here.

Reply:
I mean, it's extremely hard to follow even after that explanation, but if you've already handled it I don't feel extremely compelled to break it. Your call.